File: src/console/backend/api/editor.py
import json
from datetime import datetime, timezone
from typing import List

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from api.schemas import (
    EditorEventItem,
    EditorEventsResponse,
    FlattenCheckResponse,
    ReviewSidecarGetResponse,
    ReviewSidecarImportResponse,
    ReviewSidecarPayload,
    WorkflowStatusResponse,
)
from auth.dependencies import get_current_user
from database import get_db
from models.document import Document, DocumentStatus
from models.user import User, UserRole
from services.audit_chain_service import create_chained_audit_log
from services.editor_service import determine_editor_status

router = APIRouter(prefix="/documents", tags=["Editor Integration"])


def _require_manager(user: User) -> None:
    if user.role not in (UserRole.MANAGER, UserRole.ADMIN):
        raise HTTPException(status_code=403, detail="Manager or Admin required")


def _require_engineer(user: User) -> None:
    if user.role not in (UserRole.ENGINEER, UserRole.MANAGER, UserRole.ADMIN):
        raise HTTPException(status_code=403, detail="Engineer or higher required")


def _get_doc_or_404(doc_id: str, db: Session) -> Document:
    doc = db.query(Document).filter(Document.id == doc_id).first()
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    return doc


@router.post("/{doc_id}/review-sidecar", response_model=ReviewSidecarImportResponse)
def import_review_sidecar(
    doc_id: str,
    payload: ReviewSidecarPayload,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ReviewSidecarImportResponse:
    _require_manager(current_user)
    doc = _get_doc_or_404(doc_id, db)
    sidecar_dict = payload.model_dump()
    doc.review_sidecar = sidecar_dict
    doc.review_sidecar_imported_at = datetime.now(timezone.utc)
    doc.status = determine_editor_status(sidecar_dict)
    create_chained_audit_log(
        db,
        user_id=current_user.id,
        action="review_sidecar.imported",
        resource_type="document",
        resource_id=doc_id,
        detail=json.dumps({"schema": sidecar_dict.get("review_schema")}),
        ip_address=None,
    )
    db.refresh(doc)
    return ReviewSidecarImportResponse(
        id=doc.id,
        status=doc.status,
        review_sidecar=doc.review_sidecar,
        review_sidecar_imported_at=doc.review_sidecar_imported_at,
    )


@router.get("/{doc_id}/review-sidecar", response_model=ReviewSidecarGetResponse)
def get_review_sidecar(
    doc_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ReviewSidecarGetResponse:
    _require_engineer(current_user)
    doc = _get_doc_or_404(doc_id, db)
    return ReviewSidecarGetResponse(
        review_sidecar=doc.review_sidecar,
        review_sidecar_imported_at=doc.review_sidecar_imported_at,
    )


@router.post("/{doc_id}/flatten-check", response_model=FlattenCheckResponse)
def flatten_check(
    doc_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> FlattenCheckResponse:
    _require_manager(current_user)
    doc = _get_doc_or_404(doc_id, db)
    if doc.file_path is None:
        raise HTTPException(status_code=422, detail="Document file path is not set")
    try:
        import hashlib

        from pypdf import PdfReader

        PdfReader(doc.file_path)
        with open(doc.file_path, "rb") as f:
            raw = f.read()
        flattened_hash = hashlib.sha256(raw).hexdigest()
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=422, detail=f"Flatten check failed: {str(e)}")
    doc.status = DocumentStatus.FINALIZED
    doc.is_flattened = True
    doc.flattened_verified_at = datetime.now(timezone.utc)
    doc.flattened_hash = flattened_hash
    create_chained_audit_log(
        db,
        user_id=current_user.id,
        action="document.flatten.verified",
        resource_type="document",
        resource_id=doc_id,
        detail=json.dumps({"flattened_hash": flattened_hash}),
        ip_address=None,
    )
    db.refresh(doc)
    return FlattenCheckResponse(
        is_flattened=True,
        flattened_hash=flattened_hash,
        status=doc.status,
    )


@router.post("/{doc_id}/editor-events", response_model=EditorEventsResponse)
def post_editor_events(
    doc_id: str,
    events: List[EditorEventItem],
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> EditorEventsResponse:
    _require_engineer(current_user)
    _get_doc_or_404(doc_id, db)
    count = 0
    for event in events:
        create_chained_audit_log(
            db,
            user_id=current_user.id,
            action=event.event_type,
            resource_type="document",
            resource_id=doc_id,
            detail=json.dumps(event.detail) if event.detail is not None else None,
            ip_address=None,
        )
        count += 1
    return EditorEventsResponse(created=count)


@router.get("/{doc_id}/workflow-status", response_model=WorkflowStatusResponse)
def get_workflow_status(
    doc_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> WorkflowStatusResponse:
    _require_engineer(current_user)
    doc = _get_doc_or_404(doc_id, db)
    workflow = doc.workflow
    if workflow:
        status_str = workflow.status
        updated_at = workflow.completed_at or workflow.created_at
        steps = [
            {
                "approver_id": s.approver_id,
                "order": s.order,
                "status": s.status,
                "comment": s.comment,
                "decided_at": s.decided_at.isoformat() if s.decided_at else None,
            }
            for s in (workflow.steps or [])
        ]
    else:
        status_str = doc.status.value
        updated_at = doc.updated_at or doc.created_at
        steps = []
    editor_sync = {
        "workflow_status": status_str,
        "synced_at": datetime.now(timezone.utc).isoformat(),
    }
    doc.extra_data = {**(doc.extra_data or {}), "editor_sync": editor_sync}
    db.commit()
    db.refresh(doc)
    return WorkflowStatusResponse(
        status=status_str,
        updated_at=updated_at,
        steps=steps,
        editor_sync=editor_sync,
    )

File: src/console/frontend/src/api/editor.ts
import { api } from "./client";

// API client for the Editor integration endpoints (backend: api/editor.py +
// api/revisions.py). Types mirror api/schemas.py so the DX console can drive the
// already-implemented backend (ReviewSidecar import, flatten gate, editor
// events, workflow polling, PDF revisions).

// ─── ReviewSidecar ───────────────────────────────────────────────────────────
export interface ReviewSidecarImportResponse {
  id: string;
  status: string;
  review_sidecar: Record<string, unknown> | null;
  review_sidecar_imported_at: string | null;
}

export interface ReviewSidecarGetResponse {
  review_sidecar: Record<string, unknown> | null;
  review_sidecar_imported_at: string | null;
}

// Mirrors the CivilPDF-Editor `civilpdf.review/v1` contract (lib/review/schema.ts).
export interface ReviewSidecarPayload {
  schema?: string;
  generator?: string;
  savedAt?: string;
  stamps?: unknown[];
  annotations?: unknown[];
}

export async function getReviewSidecar(
  docId: string,
): Promise<ReviewSidecarGetResponse> {
  const res = await api.get<ReviewSidecarGetResponse>(
    `/documents/${docId}/review-sidecar`,
  );
  return res.data;
}

export async function importReviewSidecar(
  docId: string,
  payload: ReviewSidecarPayload,
): Promise<ReviewSidecarImportResponse> {
  const res = await api.post<ReviewSidecarImportResponse>(
    `/documents/${docId}/review-sidecar`,
    payload,
  );
  return res.data;
}

// ─── Flatten-check gate ──────────────────────────────────────────────────────
export interface FlattenCheckResponse {
  is_flattened: boolean;
  flattened_hash: string | null;
  status: string;
}

export async function flattenCheck(
  docId: string,
): Promise<FlattenCheckResponse> {
  const res = await api.post<FlattenCheckResponse>(
    `/documents/${docId}/flatten-check`,
  );
  return res.data;
}

// ─── Editor events (audit) ───────────────────────────────────────────────────
export type EditorEventType =
  | "stamp.placed"
  | "stamp.removed"
  | "annotation.added"
  | "comment.added";

export interface EditorEventItem {
  event_type: EditorEventType;
  detail?: unknown;
  occurred_at: string;
}

export interface EditorEventsResponse {
  created: number;
}

export async function postEditorEvents(
  docId: string,
  events: EditorEventItem[],
): Promise<EditorEventsResponse> {
  const res = await api.post<EditorEventsResponse>(
    `/documents/${docId}/editor-events`,
    events,
  );
  return res.data;
}

// ─── Workflow status (Editor polling) ────────────────────────────────────────
export interface WorkflowStep {
  approver_id: string;
  order: number;
  status: string;
  comment: string | null;
  decided_at: string | null;
}

export interface WorkflowStatusResponse {
  status: string;
  updated_at: string | null;
  steps: WorkflowStep[];
  editor_sync: Record<string, unknown> | null;
}

export async function getWorkflowStatus(
  docId: string,
): Promise<WorkflowStatusResponse> {
  const res = await api.get<WorkflowStatusResponse>(
    `/documents/${docId}/workflow-status`,
  );
  return res.data;
}

// ─── Revisions ───────────────────────────────────────────────────────────────
export interface RevisionResponse {
  id: string;
  document_id: string;
  version_number: number;
  filename: string;
  file_size: number;
  revision: string | null;
  revision_note: string | null;
  is_from_editor: boolean;
  editor_session_id: string | null;
  created_at: string;
}

export async function listRevisions(
  docId: string,
): Promise<RevisionResponse[]> {
  const res = await api.get<RevisionResponse[]>(
    `/documents/${docId}/revisions`,
  );
  return res.data;
}

export async function uploadRevision(
  docId: string,
  file: File,
  opts?: {
    revision?: string;
    revisionNote?: string;
    isFromEditor?: boolean;
    editorSessionId?: string;
  },
): Promise<RevisionResponse> {
  const form = new FormData();
  form.append("file", file);
  if (opts?.revision) form.append("revision", opts.revision);
  if (opts?.revisionNote) form.append("revision_note", opts.revisionNote);
  form.append("is_from_editor", String(opts?.isFromEditor ?? false));
  if (opts?.editorSessionId)
    form.append("editor_session_id", opts.editorSessionId);
  const res = await api.post<RevisionResponse>(
    `/documents/${docId}/revisions`,
    form,
    { headers: { "Content-Type": "multipart/form-data" } },
  );
  return res.data;
}
